### ISSUE TITLE
Find the joint eigenvalue of two matrices. 

### ISSUE THREAD

**Issue by bsun0802** [pytorch/pytorch#50204]
## 🚀 Feature
A differentiable function to find the joint eigenvalues of two matrices, MATLAB has a built-in function to find the joint eigenvalues, but I'm not sure whether it is differentiable. 
https://stackoverflow.com/questions/36551182/how-can-i-find-the-joint-eigenvalues-of-two-matrices-in-matlab

## Motivation
I encounter this problem when I want to use a distance measure for covariance matrices as the loss function, the calculation of which requires the joint eigenvalues of two matrices. 
![image](https://user-images.githubusercontent.com/39657051/103912298-12cc4380-5142-11eb-8ed1-18f5d9b618ac.png)

## Pitch
I find the function torch.lobpcg might be able to find the joint eigenvalues (but only upto k-largest) and the backward is about to break (https://github.com/pytorch/pytorch/issues/38948).
Could pytorch team implement a function to find the joint eigenvalues of two matrices which is differentiable?

## Alternatives
I'm way to layman for this level linear algebra, but I found this might be a related work-around.
https://www.alglib.net/eigen/symmetric/generalizedsymmevd.php

**Comment by zou3519** [comment]
This seems worth having. I'm curious if there is a numpy/scipy equivalent to this operation that we can use as a reference.

**Comment by mfkasim1** [comment]
I have implemented this in xitorch: https://xitorch.readthedocs.io/en/latest/api/xitorch_linalg/symeig.html#xitorch.linalg.symeig (source: https://github.com/xitorch/xitorch/blob/master/xitorch/_impls/linalg/symeig.py#L27).
Also, for scipy: https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eigh.html with `b != None`

**Comment by bsun0802** [comment]
I think by a little linear algebra, the joint eigenvalues can be calculated with native torch APIs. The process are as follows,

The joint eigenvalues are the solution of det(lambda*A - B) = 0. Since A in my case is a covariance matrix which is nonnegative positive definite, therefore A^-1 exists.

By facts, det(AB) = det(A)det(B). So det(lambdaA - B) = det(A) * det(lambda - BA^-1) = 0
And if we assume det(A) != 0, then the joint eigenvalues (lambda) are actually the eigenvalues of BA^-1.

So lambdas = torch.symeig(BA^-1) is the solution.

But torch.symeig(BA^-1) gives different answers with yours, I also tried switch the role of A and B you mentioned above, still different.

Any thoughts?

**Comment by mfkasim1** [comment]
The matrix `BA^(-1)` is not necessarily symmetric, so you have to use `torch.eig`. I have tried the method you suggested with `torch.eig`. However, in case of near-degenerate eigenvalues, the eigenvectors might be complex.

A numerically-stable solution with native torch APIs, in my opinion, is here: https://github.com/xitorch/xitorch/blob/b09d3bf449adbf39be748159f6565872ae78107e/xitorch/_impls/linalg/symeig.py#L27-L44.
I used cholesky decomposition, invert it, then use it to symmetrize the matrix, so we can still use `torch.symeig`.

**Comment by nikitaved** [comment]
You can still use symeig if you decompose
```
det(lambda A - B) = det(A^0.5 (lambda - A^{-0.5} B A^{-0.5}) A^0.5) = det(A) det(lambda * I - A^{-0.5} B A^{-0.5})
A^{-0.5} B A^{-0.5} is symmetric.
```

**Comment by bsun0802** [comment]
Thanks both of you for your time.

@mfkasim1 I think there is some lib called ALGLIB (https://www.alglib.net/eigen/symmetric/generalizedsymmevd.php) used the same cholesky decomposition method.

@nikitaved 
How to compute A^{-0.5}, is it torch.inverse(torch.matrix_power(A, 2))

**Comment by nikitaved** [comment]
@bsun0802 , no, it is a function of a matrix, not an elementwise-operation.
For a full-rank psd you can get it as
```
lambda, v = torch.symeig(pd, eigenvectors=True)
pd_sqrt = (v * lambda.sqrt().pow(-1).unsqueeze(-2)) @ v.transpose(-2, -1)
```
The approach with Cholesky decomposition could be slower because its backward will recompute `L^{-1}`. But the overall numerical stability will indeed be dependent on how well your metric matrix (`B` in your case) is conditioned.

**Comment by Lezcano** [comment]
Just a very small note, although it may or may not be useful.

In @nikitaved derivation in https://github.com/pytorch/pytorch/issues/50204#issuecomment-756788183 , you just need `A` to be symmetric and invertible and `B` PSD for `A^{-0.5} B A^{-0.5}` to be PSD (rather than needing `A` to be PSD as well). This means that it also works for general non-degenerate products.
